src/org/omegat/gui/exttrans/MachineTranslateTextArea.java

```diff
@@ -149,16 +196,25 @@ protected void setFoundResult(final SourceTextEntry se, final MachineTranslation
         UIThreadsUtil.mustBeSwingThread();
 
         if (data != null && data.result != null) {
-            displayed.add (data);
-            setText(getText() + data.result + "\n<" + data.translatorName + ">\n\n");
+            displayed.add(data);
+            StringBuilder sb = new StringBuilder();
+            for (int i = 0; i < displayed.size(); i++) {
```

Comment:
If this method can only be called sequentially on the main thread, then I guess it's OK to mutate and iterate `displayed` like this, but it makes me nervous.

Then if using `displayed` in this way is OK, it seems like it should be OK to append each incoming result to the pane instead of rebuilding the text each time.

Reply:
I think `displayed` can be sorted by `translatorName` and always display translations in same order in future.
[RFE#1467 Machine translation results sortin ](https://sourceforge.net/p/omegat/feature-requests/1467/) request this.
I'd like to try it now.

Reply:
When we move to Java 11, we can use SortedList here.

Reply:
> Then if using displayed in this way is OK, it seems like it should be OK to append each incoming result to the pane instead of rebuilding the text each time.

I'd like to try the idea adding data without rebuilding the text each time.

Reply:
It is not a really optimal and safe to inserting content by ordering. It can be like as follows changing `displayd = new TreeSet`
```java
            displayed.add(data);
            // build content
            StringBuilder sb = new StringBuilder();
            sb.append("<div id=\"").append(data.translatorName).append("\">");
            sb.append(data.result);
            sb.append("<div class=\"engine\">&lt;");
            sb.append(data.translatorName);
            sb.append("&gt;</div></div>");
            //
            HTMLDocument doc = (HTMLDocument) getDocument();
            HTMLEditorKit kit = (HTMLEditorKit) getEditorKit();
            MachineTranslationInfo previous = displayed.lower(data);
            try {
                if (previous == null) {
                    kit.insertHTML(doc, 0,  sb.toString(), 0, 0, HTML.Tag.DIV);
                } else {
                    Element el = doc.getElement(previous.translatorName);
                    kit.insertHTML(doc, el.getEndOffset(), sb.toString(), 0, 0, HTML.Tag.DIV);
                }
            } catch (BadLocationException | IOException e) {
                Log.log(e);
            }
```
But `translatorName` is not unique, because multiple Ctrl-M produce multiple results from same translator, logic is complex and difficult to handle exceptions. @amake  any suggestions?

Reply:
> If this method can only be called sequentially on the main thread, then I guess it's OK to mutate and iterate `displayed` like this, but it makes me nervous.

This method is guarded by `UIThreadsUtil.mustBeSwingThread();` Is it not enough?

Reply:
Now use `insertHTML` instead of rebuilding the text each time.
``` 
kit.insertHTML(doc, doc.getLength(), sb.toString(), 0, 0, null);
```